README.md
# Sensirion Raspberry Pi UART SFC6XXX Driver

The repository provides a driver for setting up a SFC6XXX sensor to run on a Raspberry Pi over UART using the SHDLC protocol. 

<center><img src="images/product-image-sfc6xxx.png" width="300px"></center>

Click [here](https://sensirion.com/sfc6000) to learn more about the Sensirion SFC6XXX sensor family.

## Supported sensor types

- [SFC6000](https://sensirion.com/products/catalog/SFC6000/)
- [SFC6000D-5SLM](https://sensirion.com/products/catalog/SFC6000D-5slm/)
- [SFC6000D-50SLM](https://sensirion.com/products/catalog/SFC6000D-50slm/)
- [SFC6000D-20SLM](https://sensirion.com/products/catalog/SFC6000D-20slm/)

The following instructions and examples use a *SFC6000*.


## Connect the sensor

1. Connect the sensor to a 24V power supply.
2. Connect the sensor to your Raspberry Pi by using the provided USB cable.

Please note that due to the delays introduced by the FTDI driver you can 
reach a maximum sampling frequency of about 20Hz with this setup.

For special setups check out the sensor pinout in the section below.

<details><summary>RS485 interface pinout</summary>
<p>
<img src="images/product-pinout-sfc6xxx.png" width="300px">
</p>
</details>

## Quick start example

- [Install the Raspberry Pi OS on to your Raspberry Pi](https://projects.raspberrypi.org/en/projects/raspberry-pi-setting-up)
- Download the SFC6XXX driver from [Github](https://github.com/Sensirion/raspberry-pi-uart-sfc6xxx) and extract the `.zip`
  on your Raspberry Pi
- Connect the SFC6XXX sensor as explained in the [section above](#connect-the-sensor)
- Check that the correct serial port is set in the define in `sensirion_uart_portdescriptor.h`
   - For connection over USB (in case you have other devices connected check the USB number)

     `#define SERIAL_0 "/dev/ttyUSB0"`

   - For connection over UART Pins 

     `#define SERIAL_0 "/dev/serial0"`

- Compile the driver
    1. Open a [terminal](https://projects.raspberrypi.org/en/projects/raspberry-pi-using/8)
    2. Navigate to the driver directory. E.g. `cd ~/raspberry-pi-uart-sfc6xxx`
    3. Navigate to the subdirectory example-usage.
    4. Run the `make` command to compile the driver

       Output:
       ```
       rm -f sfc6xxx_uart_example_usage
       cc -Os -Wall -fstrict-aliasing -Wstrict-aliasing=1 -Wsign-conversion -fPIC -I. -o sfc6xxx_uart_example_usage sfc6xxx_uart.h sfc6xxx_uart.c sensirion_uart_hal.h sensirion_shdlc.h sensirion_shdlc.c \ 
           sensirion_uart_hal.c sensirion_config.h sensirion_common.h sensirion_common.c sfc6xxx_uart_example_usage.c
       ```
- Test your connected sensor
    - Run `./sfc6xxx_uart_example_usage` in the same directory you used to compile the driver. You should see the
      measurement values in the console.

## Troubleshooting

### Building driver failed

If the execution of `make` in the compilation step 3 fails with something like

```bash
 make: command not found
```

your RaspberryPi likely does not have the build tools installed. Proceed as follows:

```
$ sudo apt-get update
$ sudo apt-get upgrade
$ sudo apt-get install build-essential
```



## Contributing

**Contributions are welcome!**

We develop and test this driver using our company internal tools (version
control, continuous integration, code review etc.) and automatically
synchronize the master branch with GitHub. But this doesn't mean that we don't
respond to issues or don't accept pull requests on GitHub. In fact, you're very
welcome to open issues or create pull requests :)

This Sensirion library uses
[`clang-format`](https://releases.llvm.org/download.html) to standardize the
formatting of all our `.c` and `.h` files. Make sure your contributions are
formatted accordingly:

The `-i` flag will apply the format changes to the files listed.

```bash
clang-format -i *.c *.h
```

Note that differences from this formatting will result in a failed build until
they are fixed.


## License

See [LICENSE](LICENSE).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [0.1.0] - 2023-02-08

Initial release



sfc6xxx_uart.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * Generator:    sensirion-driver-generator 0.21.0
 * Product:      sfc6xxx
 * Version:      1.1
 */
/*
 * Copyright (c) 2023, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SFC6XXX_UART_H
#define SFC6XXX_UART_H

#ifdef __cplusplus
extern "C" {
#endif

#include "sensirion_config.h"
#define SFC6XXX_SHDLC_ADDR 0x00

typedef enum {
    SET_SETPOINT_CMD_ID = 0x0,
    GET_SETPOINT_CMD_ID = 0x0,
    READ_MEASURED_VALUE_CMD_ID = 0x8,
    READ_AVERAGED_MEASURED_VALUE_CMD_ID = 0x8,
    SET_SETPOINT_AND_READ_MEASURED_VALUE_CMD_ID = 0x3,
    SET_USER_CONTROLLER_GAIN_CMD_ID = 0x22,
    GET_USER_CONTROLLER_GAIN_CMD_ID = 0x22,
    SET_USER_INIT_STEP_CMD_ID = 0x22,
    GET_USER_INIT_STEP_CMD_ID = 0x22,
    MEASURE_RAW_FLOW_CMD_ID = 0x30,
    MEASURE_RAW_THERMAL_CONDUCTIVITY_WITH_CLOSED_VALVE_CMD_ID = 0x30,
    MEASURE_TEMPERATURE_CMD_ID = 0x30,
    GET_NUMBER_OF_CALIBRATIONS_CMD_ID = 0x40,
    GET_CALIBRATION_VALIDITY_CMD_ID = 0x40,
    GET_CALIBRATION_GAS_ID_CMD_ID = 0x40,
    GET_CALIBRATION_GAS_UNIT_CMD_ID = 0x40,
    GET_CALIBRATION_FULLSCALE_CMD_ID = 0x40,
    GET_CURRENT_GAS_ID_CMD_ID = 0x44,
    GET_CURRENT_GAS_UNIT_CMD_ID = 0x44,
    GET_CURRENT_FULLSCALE_CMD_ID = 0x44,
    SET_CALIBRATION_CMD_ID = 0x45,
    GET_CALIBRATION_CMD_ID = 0x45,
    SET_CALIBRATION_VOLATILE_CMD_ID = 0x46,
    SET_SLAVE_ADDRESS_CMD_ID = 0x90,
    GET_SLAVE_ADDRESS_CMD_ID = 0x90,
    SET_BAUDRATE_CMD_ID = 0x91,
    GET_BAUDRATE_CMD_ID = 0x91,
    GET_PRODUCT_TYPE_CMD_ID = 0xd0,
    GET_PRODUCT_NAME_CMD_ID = 0xd0,
    GET_ARTICLE_CODE_CMD_ID = 0xd0,
    GET_SERIAL_NUMBER_CMD_ID = 0xd0,
    GET_VERSION_CMD_ID = 0xd1,
    DEVICE_RESET_CMD_ID = 0xd3,
} cmd_id_t;

/**
 * @brief Switch device off (close the valve by setting setpoint to 0)
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_close_valve();

/**
 * @brief sfc6xxx_set_setpoint
 *
 * Set the flow setpoint as a physical value which is used by the flow
 * controller as reference input.
 *
 * @param[in] setpoint The new setpoint.
 *
 * @note The setpoint is set to 0 if calibration is changed.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_setpoint(float setpoint);

/**
 * @brief sfc6xxx_get_setpoint
 *
 * Get the current flow setpoint as a physical value.
 *
 * @param[out] setpoint The current setpoint.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_setpoint(float* setpoint);

/**
 * @brief sfc6xxx_read_measured_value
 *
 * The command returns the latest measured flow as physical value.
 *
 * @param[out] measured_value The latest measured flow.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_read_measured_value(float* measured_value);

/**
 * @brief sfc6xxx_read_averaged_measured_value
 *
 * The command returns the average of given number of flow measurement as
 * physical value. A single measurement has a duration of 1ms, so the command
 * response time depends on the given number of measurements to average.
 *
 * @param[in] measurements number of measurements to average (1..100)
 * @param[out] averaged_measured_value The averaged flow measurement.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_read_averaged_measured_value(uint8_t measurements,
                                             float* averaged_measured_value);

/**
 * @brief sfc6xxx_set_setpoint_and_read_measured_value
 *
 * This command is a combination of the two commands set_setpoint (0x00) and
 * read_measured_value (0x08). It is intended for process data exchange
 * (setpoint and flow) and saves a lot of protocol overhead compared to separate
 * command usage.
 *
 * @param[in] setpoint The new setpoint as physical value.
 * @param[out] measured_value The latest measured flow as physical value.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_setpoint_and_read_measured_value(float setpoint,
                                                     float* measured_value);

/**
 * @brief sfc6xxx_set_user_controller_gain
 *
 * Set the user controller gain.
 *
 * @param[in] gain The user controller gain to set.
 *
 * @note This configuration is not stored in non-volatile memory of the device
 * and thus not persists after a device reset.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_user_controller_gain(float gain);

/**
 * @brief sfc6xxx_get_user_controller_gain
 *
 * Get the user controller gain.
 *
 * @param[out] gain The current user controller gain.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_user_controller_gain(float* gain);

/**
 * @brief sfc6xxx_set_user_init_step
 *
 * Set the init step of flow controller.
 *
 * @param[in] init_step the user init step to set.
 *
 * @note This configuration is not stored in non-volatile memory of the device
 * and thus not persists after a device reset.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_user_init_step(float init_step);

/**
 * @brief sfc6xxx_get_user_init_step
 *
 * Get the user init step of flow controller.
 *
 * @param[out] init_step The current user init step.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_user_init_step(float* init_step);

/**
 * @brief sfc6xxx_measure_raw_flow
 *
 * Return the measured raw flow ticks from the sensor.
 *
 * @param[out] flow Measured raw flow in ticks.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_measure_raw_flow(uint16_t* flow);

/**
 * @brief sfc6xxx_measure_raw_thermal_conductivity_with_closed_valve
 *
 * Perform a thermal conductivity measurement and return the measured raw tick
 * value. The valve is automatically closed during the measurement.
 *
 * @param[out] thermal_conductivity Measured raw thermal conductivity in ticks.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_measure_raw_thermal_conductivity_with_closed_valve(
    uint16_t* thermal_conductivity);

/**
 * @brief sfc6xxx_measure_temperature
 *
 * Return the temperature of flow sensor.
 *
 * @param[out] temperature Measured temperature [°C].
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_measure_temperature(float* temperature);

/**
 * @brief sfc6xxx_get_number_of_calibrations
 *
 * Get the number of calibrations, i.e. how many calibration the device memory
 * is able to hold.
 *
 * @param[out] number_of_calibrations Number of calibrations.
 *
 * @note Not all calibration actually contain a valid gas calibration. Use the
 * command get_calibration_validity to check which calibrations are valid and
 * thus can be activated.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_number_of_calibrations(uint32_t* number_of_calibrations);

/**
 * @brief sfc6xxx_get_calibration_validity
 *
 * Check whether there exists a valid calibration at a specific index or not.
 *
 * @param[in] index The index to check whether there is a valid calibration or
 * not.
 * @param[out] validity Whether there exists a valid calibration at the
 * specified index or not.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_calibration_validity(uint32_t index, bool* validity);

/**
 * @brief sfc6xxx_get_calibration_gas_id
 *
 * Get the gas ID of a specific calibration index.
 *
 * @param[in] index The calibration index to read the requested information
 * from.
 * @param[out] gas_id The read gas ID.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_calibration_gas_id(uint32_t index, uint32_t* gas_id);

/**
 * @brief sfc6xxx_get_calibration_gas_unit
 *
 * Get the gas unit of a specific calibration index.
 *
 * @param[in] index The calibration index to read the requested information
 * from.
 * @param[out] prefix Medium unit prefix, see datasheet for encoding.
 * @param[out] unit Medium unit, see datasheet for encoding.
 * @param[out] timebase Timebase, see datasheet for encoding.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_calibration_gas_unit(uint32_t index, int8_t* prefix,
                                         uint8_t* unit, uint8_t* timebase);

/**
 * @brief sfc6xxx_get_calibration_fullscale
 *
 * Get the fullscale flow of a specific calibration index.
 *
 * @param[in] index The calibration index to read the requested information
 * from.
 * @param[out] fullscale The read fullscale flow in the unit of the
 * corresponding calibration.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_calibration_fullscale(uint32_t index, float* fullscale);

/**
 * @brief sfc6xxx_get_current_gas_id
 *
 * Get the gas ID of the currently active calibration.
 *
 * @param[out] gas_id The read gas ID.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_current_gas_id(uint32_t* gas_id);

/**
 * @brief sfc6xxx_get_current_gas_unit
 *
 * Get the gas unit of the currently active calibration.
 *
 * @param[out] prefix Medium unit prefix, see datasheet for encoding.
 * @param[out] unit Medium unit, see datasheet for encoding.
 * @param[out] timebase Timebase, see datasheet for encoding.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_current_gas_unit(int8_t* prefix, uint8_t* unit,
                                     uint8_t* timebase);

/**
 * @brief sfc6xxx_get_current_fullscale
 *
 * Get the fullscale flow of the currently active calibration.
 *
 * @param[out] fullscale The read fullscale flow in the unit of the
 * corresponding calibration.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_current_fullscale(float* fullscale);

/**
 * @brief sfc6xxx_set_calibration
 *
 * This command stops the controller, changes the used calibration of the flow
 * controller and starts the controller. The selected calibration is stored and
 * also used after a power-on or reset.
 *
 * @param[in] calibration_number The calibration number to activate.
 *
 * @note This command will cause a write operation in a flash. Because of the
 * limited write cycles of the flash, this command should not be called
 * periodically with new calibration (max. 50’000 times). If calibration is
 * already selected the function will not do a write cycle to flash.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_calibration(uint32_t calibration_number);

/**
 * @brief sfc6xxx_get_calibration
 *
 * Get the actual set calibration number of flow controller.
 *
 * @param[out] calibration_number The current activated calibration number.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_calibration(uint32_t* calibration_number);

/**
 * @brief sfc6xxx_set_calibration_volatile
 *
 * This command stops the controller, changes the used calibration of the flow
 * controller and starts the controller. The selected calibration is not stored
 * to a non volatile memory.
 *
 * @param[in] calibration_number The calibration number to activate.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_calibration_volatile(uint32_t calibration_number);

/**
 * @brief sfc6xxx_set_slave_address
 *
 * Sets the SHDLC slave address of the device.
 *
 * @param[in] slave_address The new slave address to set.
 *
 * @note The slave address is stored in non-volatile memory of the device and
 * thus persists after a device reset. So the next time connecting to the
 * device, you have to use the new address. When changing the address of a
 * slave, make sure there isn't already a slave with that address on the same
 * bus! In that case you would get communication issues which can only be fixed
 * by disconnecting one of the slaves.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_slave_address(uint8_t slave_address);

/**
 * @brief sfc6xxx_get_slave_address
 *
 * Gets the SHDLC slave address of the device.
 *
 * @param[out] slave_address The current slave address of the device.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_slave_address(uint8_t* slave_address);

/**
 * @brief sfc6xxx_set_baudrate
 *
 * Sets the SHDLC baudrate of the device.
 *
 * @param[in] baudrate The new baudrate in bit/s. Allowed values are 9600,
 * 19200, 38400, 57600 and 115200 (default)
 *
 * @note The baudrate is stored in non-volatile memory of the device and thus
 * persists after a device reset. So the next time connecting to the device, you
 * have to use the new baudrate.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_set_baudrate(uint32_t baudrate);

/**
 * @brief sfc6xxx_get_baudrate
 *
 * Gets the SHDLC baudrate of the device.
 *
 * @param[out] baudrate Current baudrate in bit/s.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_baudrate(uint32_t* baudrate);

/**
 * @brief sfc6xxx_get_product_type
 *
 * Gets the product type from the device.
 *
 * @param[out] product_type String containing the product type.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_product_type(int8_t* product_type,
                                 uint16_t product_type_size);

/**
 * @brief sfc6xxx_get_product_name
 *
 * Gets the product name from the device.
 *
 * @param[out] product_name String containing the product name.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_product_name(int8_t* product_name,
                                 uint16_t product_name_size);

/**
 * @brief sfc6xxx_get_article_code
 *
 * Get the article code of the device. This information is also contained on the
 * product label.
 *
 * @param[out] article_code The article code as an ASCII string.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_article_code(int8_t* article_code,
                                 uint16_t article_code_size);

/**
 * @brief sfc6xxx_get_serial_number
 *
 * Gets the serial number of the SFC6xxx sensor.
 *
 * @param[out] serial_number String containing the serial number of the SFC6xxx
 * sensor.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_serial_number(int8_t* serial_number,
                                  uint16_t serial_number_size);

/**
 * @brief sfc6xxx_get_version
 *
 * Gets the version information for the hardware, firmware and SHDLC protocol.
 *
 * @param[out] firmware_major Firmware major version number.
 * @param[out] firmware_minor Firmware minor version number.
 * @param[out] firmware_debug Firmware debug state. If the debug state is set,
 * the firmware is in development.
 * @param[out] hardware_major Hardware major version number.
 * @param[out] hardware_minor Hardware minor version number.
 * @param[out] protocol_major Protocol major version number.
 * @param[out] protocol_minor Protocol minor version number.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_get_version(uint8_t* firmware_major, uint8_t* firmware_minor,
                            bool* firmware_debug, uint8_t* hardware_major,
                            uint8_t* hardware_minor, uint8_t* protocol_major,
                            uint8_t* protocol_minor);

/**
 * @brief sfc6xxx_device_reset
 *
 * Executes a reset on the device. This has the same effect as a power cycle.
 *
 * @note The device will reply before excecuting the reset.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sfc6xxx_device_reset();

#ifdef __cplusplus
}
#endif
#endif  // SFC6XXX_UART_H


sensirion_uart_hal.h
/*
 * Copyright (c) 2018, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SENSIRION_UART_HAL_H
#define SENSIRION_UART_HAL_H

#include "sensirion_config.h"
#include "sensirion_uart_portdescriptor.h"

#ifdef __cplusplus
extern "C" {
#endif

/**
 * sensirion_uart_hal_init() - initialize UART
 * 
 * @port: platform dependent port descriptor, 
 *        see sensirion_uart_typedef.h for data type
 *
 * Return:      0 on success, an error code otherwise
 */
int16_t sensirion_uart_hal_init(UartDescr port);

/**
 * sensirion_uart_hal_free() - release UART resources
 *
 * Return:      0 on success, an error code otherwise
 */
int16_t sensirion_uart_hal_free();

/**
 * sensirion_uart_hal_tx() - transmit data over UART
 *
 * @data_len:   number of bytes to send
 * @data:       data to send
 * Return:      Number of bytes sent or a negative error code
 */
int16_t sensirion_uart_hal_tx(uint16_t data_len, const uint8_t* data);

/**
 * sensirion_uart_hal_rx() - receive data over UART
 *
 * @data_len:   max number of bytes to receive
 * @data:       Memory where received data is stored
 * Return:      Number of bytes received or a negative error code
 */
int16_t sensirion_uart_hal_rx(uint16_t max_data_len, uint8_t* data);

/**
 * Sleep for a given number of microseconds. The function should delay the
 * execution for at least the given time, but may also sleep longer.
 *
 * Despite the unit, a <10 millisecond precision is sufficient.
 *
 * @param useconds the sleep time in microseconds
 */
void sensirion_uart_hal_sleep_usec(uint32_t useconds);

#ifdef __cplusplus
}
#endif

#endif /* SENSIRION_UART_HAL_H */


sensirion_uart_portdescriptor.h
/*
 * Copyright (c) 2018, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef UART_TYPEDEF_H
#define UART_TYPEDEF_H

#ifdef __cplusplus
extern "C" {
#endif

// type of uart port descriptor (platform dependent)
typedef const char* UartDescr;

// definition of default port
#define SERIAL_0 "/dev/ttyUSB0"

#ifdef __cplusplus
}
#endif
#endif

example-usage/sfc6xxx_uart_example_usage.c
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * Generator:    sensirion-driver-generator 0.21.0
 * Product:      sfc6xxx
 * Version:      1.1
 */
/*
 * Copyright (c) 2023, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include "sensirion_common.h"
#include "sensirion_uart_hal.h"
#include "sfc6xxx_uart.h"
#include <stdio.h>  // printf

#define sensirion_hal_sleep_us sensirion_uart_hal_sleep_usec

int main(void) {
    int16_t error = NO_ERROR;

    // for non linux systems modify the header file sensirion_uart_descriptor.h
    // SERIAL_0 defaults to "/dev/ttyUSB0" for linux user space implementation
    sensirion_uart_hal_init(SERIAL_0);

    error = sfc6xxx_device_reset();
    if (error != NO_ERROR) {
        printf("error executing device_reset(): %i\n", error);
        return error;
    }
    sensirion_hal_sleep_us(2000000);
    int8_t serial_number[64] = {0};
    error = sfc6xxx_get_serial_number(serial_number, 64);
    if (error != NO_ERROR) {
        printf("error executing get_serial_number(): %i\n", error);
        return error;
    }
    printf("serial_number: %s\n", serial_number);
    error = sfc6xxx_set_setpoint(2);
    if (error != NO_ERROR) {
        printf("error executing set_setpoint(): %i\n", error);
        return error;
    }
    float averaged_measured_value = 0.0;
    uint16_t repetition = 0;
    for (repetition = 0; repetition < 200; repetition++) {
        error =
            sfc6xxx_read_averaged_measured_value(50, &averaged_measured_value);
        if (error != NO_ERROR) {
            printf("error executing read_averaged_measured_value(): %i\n",
                   error);
            continue;
        }
        printf("averaged_measured_value: %.2f\n", averaged_measured_value);
    }

    error = sfc6xxx_close_valve();
    if (error != NO_ERROR) {
        return error;
    }
    return NO_ERROR;
}
